core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java

```diff
@@ -157,19 +161,21 @@ public TableBuilder withProperty(String key, String value) {
       return this;
     }
 
+    protected TableMetadata newTableMetadata() {
+      String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
```

Comment:
@ajantha-bhat, why not override `defaultWarehouseLocation`? That is already intended to let the catalog customize the table locations.

Reply:
@rdblue:  Yes, overriding  `defaultWarehouseLocation` can solve the problem of two different tables with same identifier getting the same path (issue which this PR is trying to solve)

But `defaultWarehouseLocation()` is the public interface of the catalog. 
If user calls it `defaultWarehouseLocation()` N times for a table, then it gives N different location. Which is confusing.
 
So, Instead of overrding it, I fixed as described in the PR.

Reply:
Why would we be calling default warehouse location multiple times for the same table? Sounds like we would be safe here saying that "defaultWarehouseLocation" is not deterministic.

Reply:
Because it is a user-facing API,  (nothing stops the user from calling in two places while writing custom logic around it)
I felt it looks odd to return a different result for the same input.

Reply:
@RussellSpitzer , @rdblue : I have replied above that why I didn't modify `defaultWarehouseLocation() `, If you still feel it is better to override `defaultWarehouseLocation() ` please let me know. I just want this issue to be fixed. Thanks.